core/module.go

```diff
@@ -282,39 +282,34 @@ func (mod *Module) ModTypeFor(ctx context.Context, typeDef *TypeDef, checkDirect
 
 	switch typeDef.Kind {
 	case TypeDefKindString, TypeDefKindInteger, TypeDefKindBoolean, TypeDefKindVoid:
-		modType, ok = mod.modTypeForPrimitive(ctx, typeDef, checkDirectDeps)
+		modType, ok = mod.modTypeForPrimitive(typeDef)
 	case TypeDefKindList:
 		modType, ok, err = mod.modTypeForList(ctx, typeDef, checkDirectDeps)
 	case TypeDefKindObject:
-		modType, ok, err = mod.modTypeForObject(ctx, typeDef, checkDirectDeps)
-		if checkDirectDeps && ok {
-			return modType, true, nil
+		modType, ok, err = mod.modTypeFromDeps(ctx, typeDef, checkDirectDeps)
+		if ok || err != nil {
+			return modType, ok, err
 		}
-
+		modType, ok = mod.modTypeForObject(typeDef)
 	case TypeDefKindInterface:
-		// For some reason, if we do this logic inside `modTypeForInferface`, it doesn't work...
-		if checkDirectDeps {
-			// check to see if this is from a *direct* dependency
-			depType, ok, err := mod.Deps.ModTypeFor(ctx, typeDef)
-			if err != nil {
-				return nil, false, fmt.Errorf("failed to get interface type from dependency: %w", err)
-			}
-			if ok {
-				return depType, true, nil
-			}
+		modType, ok, err = mod.modTypeFromDeps(ctx, typeDef, checkDirectDeps)
+		if ok || err != nil {
+			return modType, ok, err
 		}
-
-		modType, ok = mod.modTypeForInterface(ctx, typeDef, checkDirectDeps)
+		modType, ok = mod.modTypeForInterface(typeDef)
 	case TypeDefKindScalar:
-		modType, ok, err = mod.modTypeForScalar(ctx, typeDef, checkDirectDeps)
-		if checkDirectDeps && ok {
-			return modType, true, nil
+		modType, ok, err = mod.modTypeFromDeps(ctx, typeDef, checkDirectDeps)
+		if ok || err != nil {
+			return modType, ok, err
 		}
+		modType, ok = nil, false
+		slog.ExtraDebug("module did not find scalar", "mod", mod.Name(), "scalar", typeDef.AsScalar.Value.Name)
 	case TypeDefKindEnum:
-		modType, ok, err = mod.modTypeForEnum(ctx, typeDef, checkDirectDeps)
-		if checkDirectDeps && ok {
-			return modType, true, nil
+		modType, ok, err = mod.modTypeFromDeps(ctx, typeDef, checkDirectDeps)
+		if ok || err != nil {
+			return modType, ok, err
 		}
+		modType, ok = mod.modTypeForEnum(typeDef)
```

Comment:
Nothingburger, preserved for posterity:

Does this mean we'll find a dependency's definition before our own?

Actually, it looks like we did that before, too.

Is that weird? 🤔 Maybe a pre-existing bug that just hasn't come up yet? Or maybe it just doesn't matter, because TypeDefs are namespaced. That must be it. There's no "shadowing", your TypeDef is `Foo` and their TypeDef is `SomeModule_Foo`, so we'd never find it in both.

I guess we could even do something silly like "if it has a `_`, only look in deps, and if it doesn't, don't look in deps."

Reply:
> ? Or maybe it just doesn't matter, because TypeDefs are namespaced. That must be it.

Yeah, this is it - all items are namespaced, including the "self" module. I actually expect that if we had a name clash we'd have more pressing issues other than "dependencies get resolved first", I think there'd be some fun graphql funkiness in there, we expect unique names all through there.